compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp

```diff
@@ -405,19 +408,133 @@ CollapseInfo::getTransformedCollapsableLoops(AffineMap map) const {
   }
-  collapsableLoops.set_subtract(toRemove);
-  updateReassociation();
-  return toRemove.size();
+  return transformedReassociation;
+}
+
+// Update CollapseInfo to ensure that all dimensions collapsable in `this` are
+// also collapsable in `otherInfo`. This means:
+// 1. Any dimension not collapsable in `otherInfo` should not be collapsable in
+// `this`
+// 2. For any pair of dimensions in `this`, if they are collapsable in
+// `consumerInfo`, they must be collapsable into the same dimension in
+// `consumerInfo` to be collapsable into the same dimension in `this`.
+bool CollapseInfo::updateFromConsumer(OpOperand *operand,
+                                      const CollapseInfo &consumerInfo) {
+  bool didChange = false;
+  FailureOr<AffineMap> consumerToProducerMap =
+      getConsumerLoopToProducerLoopsMap(*operand);
+  if (failed(consumerToProducerMap)) {
+    this->clear();
+    return !consumerInfo.getCollapsibleLoops().empty();
+  }
+
+  FailureOr<CollapsableLoopsSet> consumerCollapsable =
+      consumerInfo.getTransformedCollapsableLoops(
+          consumerToProducerMap.value());
+  if (failed(consumerCollapsable)) {
+    didChange = collapsableLoops.size();
+    this->clear();
+    return didChange;
+  }
+
+  FailureOr<SmallVector<ReassociationIndices>> consumerReassoc =
+      consumerInfo.getTransformedReassociation(consumerToProducerMap.value());
+  assert(succeeded(consumerReassoc));
+
+  // Get a map from original index to the index it gets collapsed into
+  llvm::DenseMap<long, long> consumerCollapseMap;
+  for (const auto &[idx, indicies] : llvm::enumerate(consumerReassoc.value())) {
+    for (const auto elem : indicies) {
+      consumerCollapseMap[elem] = idx;
+    }
+  }
+
+  // Remove all collapsable loops in `producer` that are not collapsable in
+  // `consumer` (set intersect)
+  didChange = collapsableLoops.remove_if([&](long elem) -> bool {
+    return !consumerCollapsable.value().contains(elem);
+  });
+
+  // Now update the reassociation indicies given the updated `collapsableLoops`
+  // and `consumerCollapsableMap`.
+  // The idea is to reconstruct the reassociation indicies, and at each index:
+  // (1) If `index` IS NOT in `collapsableLoops`, split `indicies` and don't add
+  // `index` to either.
+  //
+  // (2) If `index` IS in `collapsableLoops` but `consumerCollapseMap` maps
+  // `index` to a different collapsed loop then the other indicies,  split
+  // `indicies` and insert `index` into the new one.
+  //
+  // For example:
+  // producer reassociation = [[0, 1], [2, 3]]
+  // consumer reassociation = [0, 1, 2, 3]
+  // then, consumer reassociation gets updated to [[0, 1], [2, 3]] because
+  // [0, 1] and [2, 3] get collapsed into different loops
+  //
+  // (3) Otherwise, keep the index
+  SmallVector<ReassociationIndices> newReassociation;
+  for (ReassociationIndicesRef indicies : reassociation) {
+    // Track the loop index that `indicies` get collapsed into.
+    long collapseIntoIdx = -1;
```

Comment:
nit: declare a constant variable to avoid magic number. This also helps readability. E.g.,

```cpp
constexpr long kUninitialized = -1; // And use the variable in the below code.
long collapseIntoIdx = kUninitalized;
```